### PR TITLE
Expose symmetric encryptions

### PR DESCRIPTION
**Pull request by samtay**
On the outside, all this does is expose extra `encrypt_symmetric_*` methods wherever there were `encrypt_*` methods, on both the `Runtime` and the `LogProofBuilder`. Internally I've changed a few things
- Gotten rid of the `runtime::BfvEncryptionComponents`. This lived at the runtime level and contained the higher level plaintext and ciphertext types that are supposed to be agnostic to seal, however it also contained types specific to seal. And it kind of forced a wrapping and unwrapping of seal data, when logproof itself is already seal specific. So instead, just define and deal with types `AsymmetricComponents` and `SymmetricComponents` from the seal_fhe layer. The builder interacts with these as the runtime iterates over the inner plaintexts.
- Added a marker to the `seal_fhe::Encryptor` to track whether or not it supports sym/asym encryptions.

See https://github.com/Sunscreen-tech/SEAL/pull/8 for seal changes.

### CODE REVIEW

File: seal_fhe/src/encryptor_decryptor.rs [hunk 1]
```diff
@@ -35,10 +79,48 @@ use crate::{Ciphertext, Context, Plaintext, PublicKey, SecretKey};
...
+pub struct Encryptor<T = ()> {
     handle: *mut c_void,
+    _marker: PhantomData<T>,
+}
+
+/// An encryptor capable of symmetric encryptions.
+pub type SymmetricEncryptor = Encryptor<Sym>;
+
+/// An encryptor capable of asymmetric encryptions.
+pub type AsymmetricEncryptor = Encryptor<Asym>;
```
Comment (ryanorendorff) [inline seal_fhe/src/encryptor_decryptor.rs:91]: Why not have a type for `Encrytor<SymAsym>`?.

File: sunscreen_runtime/src/builder.rs [hunk 2]
```diff
@@ -409,44 +477,89 @@ mod linked {
...
         ) -> Result<Ciphertext> {
-            let enc_components = self.runtime.encrypt_return_components_switched_internal(
-                message.pt(),
-                message.type_name(),
-                public_key,
-                true,
-                None,
-            )?;
             let existing_idx = message.linked_id();
+            let mut i = 0;
```
Comment (ryanorendorff) [inline sunscreen_runtime/src/builder.rs:504]: Similarly we could consider renaming `encrypt_internal` with `encrypt_asymmetric_internal`.

**Review by ryanorendorff — APPROVED** [review]
Looks good! The cleanup of the components and splitting into the AsymmetricComponents and SymmetricComponents is a nice abstraction.

File: logproof/src/bfv_statement.rs [hunk 3]
```diff
@@ -792,23 +795,28 @@ mod tests {
...
+            let num_public_statements = rng.gen_range(1..=3);
+            let num_private_statements = rng.gen_range(1..=3);
+            let num_duplicate_public_msgs = rng.gen_range(0..num_public_statements);
+            let num_duplicate_private_msgs = rng.gen_range(0..num_private_statements);
+            self.random_fixture_with(
+                num_public_statements,
+                num_private_statements,
+                num_duplicate_public_msgs,
+                num_duplicate_private_msgs,
+            )
```
[the remark this answers is not in the document]
Comment (samtay) [reply logproof/src/bfv_statement.rs:807]: I think I'll be able to use it after my next PR

File: logproof/src/bfv_statement.rs [hunk 4]
```diff
@@ -823,32 +831,62 @@ mod tests {
 
             // statements without duplicate messages
             for (i, m) in messages.iter().enumerate() {
-                let (ct, u, e, r) = self
+                if i < num_unique_public_statements {
+                    let (ct, components) = self
+                        .encryptor
+                        .encrypt_return_components(&m.plaintext)
```
[the remark this answers is not in the document]
Comment (samtay) [reply logproof/src/bfv_statement.rs:837]: I guess I'm not opposed to that for this particular method, which exists on the `seal_fhe::Encryptor` and is not part of the public API. However I wouldn't want to change the names of the `pub` methods on the runtime, because for most users it is simpler to not really consider asym vs sym and just default to asym.

File: seal_fhe/src/encryptor_decryptor.rs (hunk 1 above)
Comment (samtay) [reply seal_fhe/src/encryptor_decryptor.rs:91]: Honestly because I thought `SymmetricAsymmetricEncryptor` was too wordy :shrug: I only ended up needing to reference `Encryptor<SymAsym>` in a couple tests so I thought it was fine to not make a synonym for it. If you want one I can add it.

File: logproof/src/bfv_statement.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryanorendorff) [reply logproof/src/bfv_statement.rs:807]: Sounds good!

File: logproof/src/bfv_statement.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ryanorendorff) [reply logproof/src/bfv_statement.rs:837]: Hmm that's fair. The asymmetry (har har) is a tad odd but I agree that it doesn't make sense to change the public names.

File: seal_fhe/src/encryptor_decryptor.rs (hunk 1 above)
Comment (ryanorendorff) [reply seal_fhe/src/encryptor_decryptor.rs:91]: Might be worth adding just for completeness.
